### PR TITLE
[Tutorial] tutorial to writing a costumized pass

### PR DESCRIPTION
**Pull request by were**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by tqchen** [comment]
please request reviews

**Comment by tqchen** [comment]
@nishi-t @eqy  please review

File: tutorials/language/pass.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    if isinstance(op, tvm.stmt.For):
+        if isinstance(op.extent, tvm.expr.IntImm):
+            if op.extent.value % 8 == 0:
+                loops.append(op)
+
+#####################################################################
+# IR Transformation
+# ~~~~~~~~~~~~~~~~~
+# Transformation interface is slightly different from the visitor interface. There is only
+# post-order callback in the visitor, but transformation visitor supports both pre-order and
```
[the remark this answers is not in the document]
Comment (eqy) [reply tutorials/language/pass.py]: "both a pre-order and"

File: tutorials/language/pass.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+# post-order callback. If you want to keep the origin IR node, just return None. If you want to
```
Comment (eqy) [inline tutorials/language/pass.py]: "a post-order callback."

File: tutorials/language/pass.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+# change the current node to some node you desire, use TVM IR maker interface to build it and return
```
Comment (eqy) [inline tutorials/language/pass.py]: "change the current node to some other node, ..."

File: tutorials/language/pass.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    # Thus, in this case only `For` nodes will call `vectorize8`
+    stmt = tvm.ir_pass.IRTransform(stmt, None, vectorize8, ['For'])
+    
+    return stmt
+
+#####################################################################
+# Glue to Lowering
+# ----------------
+# So far, we are done with writing this IR transformation pass. What we need to do next is to glue
+# this pass to TVM's lower pass. We can first call this function directly to give us some sanity.
```
Comment (eqy) [inline tutorials/language/pass.py]: "We can first call this function directly as a sanity check."

File: tutorials/language/pass.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+#
+
+print(vectorize(ir))
+
+#####################################################################
+# In TVM, there is a property called ``BuildConfig``. You can use this property to customize your
+# own lowering options. In this case, we inject a pass written above into the TVM standard lower
```
Comment (eqy) [inline tutorials/language/pass.py]: "the pass written above... into the TVM standard lowering"

File: tutorials/language/pass.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+# pass by feeding **a list of tuple** to argument ``add_lower_pass``. "Tuple" indicates different
```
Comment (eqy) [inline tutorials/language/pass.py]: "as the argument to ``add_lower_pass``."

File: tutorials/language/pass.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+# phases of lowering. In TVM, there are four phases lowering and customized ones will be called
```
Comment (eqy) [inline tutorials/language/pass.py]: "four phases of lowering"

File: tutorials/language/pass.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+# after each phase.
+#
+# .. note::
+#     Here are the essential contributions made by each phase:
+#     Phase 0 generates the raw IR and loop levels; phase 1 flattens the storage; phase 2 transforms
+#     the loop levels, like unroll, vectorization and threading; phase 3 does some cleanup work.
```
Comment (eqy) [inline tutorials/language/pass.py]: We should not link together four independent clauses with just semicolons.

File: tutorials/language/pass.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+#
+# Thus, we believe after phase 1 is a good place to put this transformation pass.
```
Comment (eqy) [inline tutorials/language/pass.py]: "A good place to put this transformation pass is just after phase 1."

File: tutorials/language/pass.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+#
+
+with tvm.build_config(add_lower_pass=[(1, vectorize)]) as cfg:
+    print(tvm.lower(sch, [a, b, c], simple_mode=True))
+
+#####################################################################
+# Summary
+# -------
+# This tutorial gives a quick view of writing a customized IR transformation pass:
```
Comment (eqy) [inline tutorials/language/pass.py]: "quick overview"

File: tutorials/language/pass.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+# - Use ``tvm.ir_pass.PostOrderVisit`` to gather information of each IR nodes.
```
Comment (eqy) [inline tutorials/language/pass.py]: "on each IR node."

File: tutorials/language/pass.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+# - Use ``tvm.ir_pass.IRTransform`` to transform IR nodes.
+# - Wrap up two above to write a ``stmt->stmt`` function.
```
Comment (eqy) [inline tutorials/language/pass.py]: "wrap the two above in a ? ``stmt->stmt`` ..."

File: tutorials/language/pass.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+# - Use ``tvm.build_config`` to glue this function to TVM lower pass
```
Comment (eqy) [inline tutorials/language/pass.py]: "to a TVM lower pass."

File: tutorials/language/pass.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+######################################################################
+# Writing a Pass
+# --------------
+# Essentially, an "IR transformation pass" is a function which maps a statement to a new statement.
+# Thus, we define this vectorize function and implement it step by step.
+#
+
+######################################################################
+# TVM already provides two class for users to both analyze the IR and transform IR.
```
[the remark this answers is not in the document]
Comment (were) [reply tutorials/language/pass.py]: I think "analyze" should be before "transform"

File: tutorials/language/pass.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+=========================
+**Author**: `Jian Weng <https://were.github.io>`_
+
+TVM is a framework to abstract the heterogenity of those various machine learning
+accelerators. Sometimes users may want to customized some analysis and IR transformation
+to adopt TVM to their own specialized hardware. This tutorial helps users write
+a customized pass in TVM.
+ Prerequisites
+-------------
+Before reading this tutorial, we assume readers have already known these well:
```
[the remark this answers is not in the document]
Comment (eqy) [reply tutorials/language/pass.py]: We can write "we assume readers already know these topics well"

File: tutorials/language/pass.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (eqy) [reply tutorials/language/pass.py]: Yes, +1

**Comment by xqdan** [comment]
How can I do MergeSeq or MergeNest in python, should we expose these api to python?

**Comment by were** [comment]
@xqdan Can you make the question more specific?
At the first glance I even suppose this question is not related at all.

**Comment by were** [comment]
@eqy I resolved your requests!

**Comment by were** [comment]
@xqdan Can you be more specific? What util (I suppose ur 'until' is typo) do you need?

**Comment by xqdan** [comment]
@were I mean we may need more utility functions when writing python pass, eg. MergeNest in c++,
https://github.com/dmlc/tvm/blob/b95b5958913927b90463dddb61eb18ef6e1556f6/src/pass/ir_util.cc#L11

**Comment by were** [comment]
@xqdan but it is not `stmt->stmt` mapping at all.

**Comment by were** [comment]
also, i believe this c++ code can be done in equivalent python in tvm

**Comment by xqdan** [comment]
@were consider you want to insert two new stmts into a stmt before, and one new stmt after, how do you do that? that's usual requirement when you writing a pass.

**Comment by xqdan** [comment]
@were actually it's stmt -> stmt, because stmt can be big, can be small, the whole AST can be a stmt, which also can be divided into small ones

**Comment by were** [comment]
@tqchen My changes have been approved.

**Comment by were** [comment]
@yzhliu I have done.

File: tutorials/language/pass.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
+"""
```
Comment (tqchen) [inline tutorials/language/pass.py]: pass -> low_level_custom_pass.py

File: tutorials/language/pass.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+Writing a Customized Pass
+=========================
```
Comment (tqchen) [inline tutorials/language/pass.py]: Let us put it under tutorials/dev

**Comment by were** [comment]
@tqchen Done.

**Comment by tqchen** [comment]
need to update ordering here 
https://github.com/dmlc/tvm/blob/master/docs/conf.py#L196
and add a file like https://github.com/dmlc/tvm/blob/master/tutorials/autotvm/README.txt

**Comment by were** [comment]
@tqchen Done.

**Comment by tqchen** [comment]
Thanks @were @eqy @xqdan  , this is now merged.

**Pull request merged**